src/polus/plugins/_plugins/classes/plugin_methods.py

```diff
@@ -187,6 +213,83 @@ def __setattr__(self, name, value):
 
         super().__setattr__(name, value)
 
+    def _to_cwl(self):
+        """Return CWL yml as dict."""
+        cwl_dict = CWL_BASE_DICT
+        cwl_dict["inputs"] = {}
+        cwl_dict["outputs"] = {}
+        inputs = [input_to_cwl(x) for x in self.inputs]
+        inputs = inputs + [output_to_cwl(x) for x in self.outputs]
+        for inp in inputs:
+            cwl_dict["inputs"].update(inp)
+        outputs = [outputs_cwl(x) for x in self.outputs]
+        for out in outputs:
+            cwl_dict["outputs"].update(out)
+        cwl_dict["requirements"]["DockerRequirement"]["dockerPull"] = self.containerId
```

Comment:
This method implementation is couple to the current "template" approach to generate CWL. It also only work when the Command Line Tools executables are docker containers.

Reply:
https://www.commonwl.org/user_guide/faq.html#use-cwl-and-cwltool-with-singularity FYI "Directly specifying a Singularity format container is not part of the CWL standards."

Reply:
Docker format containers can be automatically converted to Singularity containers to run on HPC machines using cwl-utils https://github.com/common-workflow-language/cwl-utils  If you really need to directly run a Singularity image and don't mind relying on implementation details / violating the spec, it can be done.